lib/tropic/tropic_paragraph.flow

```diff
@@ -1142,175 +1137,109 @@ reflowTParaWordsRec(
-				} else {
-					detachWord(word, true);
-					reflowTParaWordsRec(
-						rest,
-						nextWordIndex,
-						nextLineWidth,
-						nextLineWidth,
-						[],
-						addLine(currentWords, lineIndent),
-						paraIndent,
-						false,
-						catchOverflow,
-						ignoreLetterSpacing
-					);
-				};
-			}
-			GeneralDynamicSpace(__, __, __) : {
-				width = getWordWidth(word);
-				if (width < remaining || !isSpaceWord(word)) {
-					reflowTParaWordsRec(
-						rest,
-						nextWordIndex,
-						availableWidth,
-						remaining - width,
-						arrayPush(currentWords, word),
-						result,
-						paraIndent,
-						false,
-						catchOverflow,
-						ignoreLetterSpacing
-					);
-				} else {
-					detachWord(word, true);
-					reflowTParaWordsRec(
-						rest,
-						nextWordIndex,
-						nextLineWidth,
-						nextLineWidth,
-						[],
-						addLine(currentWords, lineIndent),
-						paraIndent,
-						false,
-						catchOverflow,
-						ignoreLetterSpacing
-					);
-				}
-			};
-			default: {
+			GeneralSpace(__, __, __) : spaceCaseFn();
+			GeneralDynamicSpace(__, __, __) : spaceCaseFn();
+			default : {
 				width = ref 0.0;
+				widthWithBreakEquation = ref 0.0;
+				wordsCountWithBreakEquation = ref 0;
 				getWordView = \w -> w.ghostView;
 
-				form = getWordView(word);
-				if (remaining == 31.) {
-					println("");
-				}
 				// Number of words that should be on the same line
-				wordsCount = iteriUntil(words, \i, _word : TParaWord -> {
+				wordsCount0 = iteriUntil(words, \i, _word : TParaWord -> {
 					addWidth = \-> width := ^width + getWordWidth(_word);
 
-					spaceCase = \str -> if(isSpace(str)) {
+					spaceCase = \str -> if (isSpace(str)) {
 						true
 					} else {
 						addWidth();
 						false;
 					};
 
 					switch (_word.word) {
-						NewLine(): true;
-						GeneralSpace(str, __, __): spaceCase(str);
-						GeneralDynamicSpace(str, __, __): spaceCase(str);
-						GeneralLinePart(first, mid, end, allowBreakAfter): {
+						NewLine() : true;
+						GeneralSpace(str, __, __) : spaceCase(str);
+						GeneralDynamicSpace(str, __, __) : spaceCase(str);
+						GeneralLinePart(first, mid, end, allowBreakAfter) : {
 							evaluatePositionForGeneralLinePart(i, _word);
 							addWidth();
 							allowBreakAfter;
 						}
-						default: {
+						default : {
 							f = getWordView(_word);
 							// it is no use to make line with zero-width content, so we continue counting if the width is 0
 							// For example, single WigiRecursive in line get surrounded by WigiText("",[]), WigiRecursive, WigiText("",[])
 							// such construction becomes [Empty, form, Empty] and we should keep them on the same line
+							previousWord = elementAt(words, i - 1, dummyTParaWord);
+							keepTogether = keepWordsTogether(f, _word, previousWord, true);
 							if (
-								(isMarked && i <= 1) ||
-								keepWordsTogether(f, _word, if (i < 1) dummyTParaWord else words[i - 1]) ||
-								^width == 0.0
+								isMarked && i <= 1 ||
```

Comment:
this code looks odd, not sure what exactly it should do and why it is so complicated